### PR TITLE
Added margo_set_progress_when_needed

### PR DESCRIPTION
**Pull request by mdorier**

### CODE REVIEW

File: src/margo-instance.h [hunk 1]
```diff
@@ -196,6 +205,50 @@ typedef struct {
...
+        if ((__mid__)->progress_when_needed.flag) {       \
+            ABT_mutex_lock(ABT_MUTEX_MEMORY_GET_HANDLE(   \
+                &(__mid__)->progress_when_needed.mutex)); \
+            --(__mid__)->progress_when_needed.pending;    \
+            ABT_mutex_unlock(ABT_MUTEX_MEMORY_GET_HANDLE( \
+                &(__mid__)->progress_when_needed.mutex)); \
+        }                                                 \
+    } while (0)
+
+#define WAIT_FOR_PROGRESS_TO_BE_NEEDED(__mid__)                             \
```
Comment (carns) [inline src/margo-instance.h:235]: Maybe these macros can be disabled (i.e. no locking etc.) if margo is initialized in listening mode?  In that case (on the server side, when unexpected requests could arrive at any time) we always have to call progress, but on clients you only need to call progress when there is margo activity.

**Comment by mdorier** [comment]
It's disabled by default unless you call `margo_set_progress_when_needed(mid, true)`. I don't want to disable it automatically when the `mid` is listening, because for instance in Mofka the client has to be listening: it sends RPCs to the server and within these RPCs, the server sends RPCs back to the client. In this situation the fact that the client has an on-going RPC makes it such that the progress loop is unblocked and ready to receive the RPC from the server.

**Comment by mdorier** [comment]
Experiments with Mofka shows that it works as intended. @carns any objection to merging it? Should I make this behavior the default for margo instances initialized with listening = false? (technically I think you could have situation where things break, e.g. the client sends an RPC with a bulk handle to the server, the server responds, and keeps the bulk handle around to make RDMA operations later, outside of the RPC, and you're using a tcp transport or any transport where RDMA is emulated and requires explicit progress).

**Comment by carns** [comment]
Oh, I see.  I overlooked the logic was opt-in.  Also I didn't realize that mofka clients were receiving RPCs, that's interesting.

At any rate, I'm fine with this being merged, but my preference would be for it to remain opt in for now.  Possibly automatically enabled later but we should take our time thinking about the use cases.

**Comment by mdorier** [comment]
Ok, merging.

**Pull request merged**